IotHubConsumer/Receiver.cs

```diff
@@ -39,18 +42,7 @@ await foreach (var partitionEvent in consumer.ReadEventsAsync(cancellationToken)
                     var data = Encoding.UTF8.GetString(partitionEvent.Data.Body.ToArray());
                     Console.WriteLine("\t{0}:", data);
 
-                    var retainFlag = "false";
-                    if (partitionEvent.Data.Properties.ContainsKey("mqtt-retain"))
-                    {
-                        retainFlag = partitionEvent.Data.Properties["mqtt-retain"].ToString();
-                    }
-
-                    var messageType = "telemetry";
-                    if (partitionEvent.Data.Properties.ContainsKey("iothub-MessageType"))
-                    {
-                        messageType = partitionEvent.Data.Properties["iothub-MessageType"].ToString();
-                    }
-                    receivedMessages.Add(new D2CMessage { Payload = data, RetainFlag = retainFlag, MessageType = messageType });
+                    receivedMessages.TryAdd(data, partitionEvent.Data.Properties);
```

Comment:
TryAdd is here because each partition sends a message. Previously we just added each one into the hashset, because each object would be hashed based on their memory address, which would be different.

I also believe that hashset wasn't O1 due to the above statement.

Reply:
I think it is ok that you use TryAdd here, although if each partition sends a message and data is never null, we can also use Add.

However, the expected complexity of search operation in a HashSet is constant - O(1), even if you populate with user defined types.

"A HashSet holds a set of objects, but in a way that it allows you to easily and quickly determine whether an object is already in the set or not. It does so by internally managing an array and storing the object using an index which is calculated from the hashcode of the object. 

HashSet is an unordered collection containing unique elements. It has the standard collection operations Add, Remove, Contains, but since it uses a hash-based implementation, these operations are O(1). (As opposed to List for example, which is O(n) for Contains and Remove.) "

This is taken from this link 
https://stackoverflow.com/questions/4558754/define-what-is-a-hashset#:~:text=A%20HashSet%20holds%20a%20set%20of%20objects%2C%20but,hashcode%20of%20the%20object.%20Take%20a%20look%20here

Reply:
> I think it is ok that you use TryAdd here, although if each partition sends a message and data is never null, we can also use Add.

`Add` crashes on the second partition with a key conflict. Meanwhile the hashset added multiple copies of the object

> [a hashset] has the standard collection operations Add, Remove, Contains [...] these operations are O(1)

True, but `FirstOrDefault` is neither, it is _Find / Search_. We don't have an object to check if it exists in the set. Instead, we're looking through all elements to see if one matches the condition. We need key-value pairs to have O(1) search.

We did a bit of experimentation on the above here: https://github.com/fhammerl/HashSetByProp/blob/master/HashSetByProp/Program.cs